feat: add formatting for errors to make multiline stacktraces in helm templates

jessesimpson36 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:

closes https://github.com/helm/helm/issues/13447

The purpose of this PR is to bring some structure into the error messages that helm will output. This structure is necessary so that readers of these error messages will naturally be able to identify important details like where the error occurred, and what the error actually is. For me, this problem comes about when you have complicated logic inside _helpers files, and is made worse when those _helpers call other _helpers.

This PR addresses this problem by adding whitespace to parts of the error message.  The structure of this format is : 
```
fileLocation:lineNumber:colNumber
  functionCallWithinFile
    contentOfError
```
and in the unit test I added, a real world example would look like this:
```
        NestedHelperFunctions/templates/svc.yaml:1:9
          executing "NestedHelperFunctions/templates/svc.yaml" at <include "nested_helper.name" .>: 
            error calling include:
        NestedHelperFunctions/templates/_helpers_1.tpl:1:39
          executing "nested_helper.name" at <include "common.names.get_name" .>: 
            error calling include:
        NestedHelperFunctions/charts/common/templates/_helpers_2.tpl:1:50
          executing "common.names.get_name" at <.Release.Name>: 
            nil pointer evaluating interface {}.Name
```

This change may introduce some risk that any added formatting logic may delete important parts of the error message.  As an attempt to address this, I use regex's to match against the known error messages from the parent golang text/template library, and if there is any part of the error that doesn't match those regex's,  the unformatted error message will be returned.


**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

jessesimpson36 committed:
wip: draft at making cleaner stacktraces

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
fix: adjust test to not require external chart

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
fix: split up the multiline errors to be more vertical

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
style: create String function for TraceableError representing one chunk of a stacktrace

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
style: use interface functions instead of inline logic

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
style: removed variable only used once

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
style: consolidate for loops

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
fix: save to fileLocation rather than fileLocations[i] which gets overwritten

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
style: renamed i variable for consistency

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
fix: remove comparison from old error message to new

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
fix: add quality checks to ensure formatted error doesnt remove important info

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
test: adjusted to make it more meaningful and to pass

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
fix: added protection against while-true condition in unbounded for loop

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

robertsirc commented:
Whats up yo. Thanks for the PR, we don't want to back port these changes but these are good for Helm 4. You do have errors that need to be addressed.

jessesimpson36 committed:
test: adjust formatting for error in test

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
style: make format

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 commented:
I added a commit to fix one test that was failing and to run `make format`. However, there are still some tests failing. I'll take some time to address later today.

jessesimpson36 committed:
fix: add some index checking and fixed a test that relied on type-checking

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 commented:
Tests and linting should pass now. For the test that relied on `err.(type)`, I currently don't know how to handle it to preserve the type checking.  I'm open to suggestions.

gjenkins8 reviewed (commented):
Thanks for the PR! Yes, better error messages would be welcome. The concern I think for Helm is the string parsing of the upstream go text/template error messages potentially being brittle.

The first thought, is did we ask upstream to improve their error type? So Helm could potentially format errors strings directly, rather than trying to parse them? (and e.g. reference this PR for why it would be nice for text/template to improve error messages).

Any upstream change will take time (presumably). In the interim We can look into text/templates's code. And see what it might do with respect to errors. A quick look and it seems like there are three places it formats errors upon execution. Lines 138, 141, 191

https://cs.opensource.google/go/go/+/refs/tags/go1.23.6:src/text/template/exec.go;l=114-147

Can we write regex's to specifically match these strings? And extract the parameters? The PR has a lot of string manipulation. I suspect if we can simple extract and format, a lot of logic will become unnecessary. And a set of specific regexes will be less brittle (ie. if none match, we would return the original error). Thoughts?

jessesimpson36 commented:
These are good ideas. I'm willing to explore what it would take to upstream the formatting changes in text/template. and using regex's  also could be good at reducing the total line's added on this patch. It may take a few weeks for me to get the time to make these changes / start a conversation with the contributors of text/template. 

> Why not just ?
> 
```diff
-	return fmt.Errorf("%s", originalErr.Error())
+	return originalErr
```
:laughing: yeah that line looks silly now.

jessesimpson36 commented:
Here is my issue proposing additional fields to the ExecError type. https://github.com/golang/go/issues/71801

jessesimpson36 committed:
fix: use originalErr instead of formatting original error

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
refactor: make use of regexs for err parsing

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 commented:
@gjenkins8 , I added a commit replacing much of the string manipulation logic with regex's. I also removed the logic that compares whether the old error message and the new are similar enough,  because if the regex's do not match, then the old error message will be returned anyway.

What do you think?

jessesimpson36 committed:
test: adjust failing test on extra whitespace

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 commented on pkg/engine/engine.go:
```diff
@@ -321,6 +321,74 @@ func cleanupParseError(filename string, err error) error {
+	if equivalenceRating >= 80 {
+		return formattedErr
+	}
+	return fmt.Errorf("%s", originalErr.Error())
```
I made this modification, but also in the following commit, this function became unnecessary. Marking as resolved.

jessesimpson36 committed:
refactor: remove more unnecessary format calls

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 commented:
I'm sorry, I asked for a re-review and I'm still making refactoring changes. I'll update here when I'm done

jessesimpson36 committed:
refactor: prevent duplicates being inserted rather than post-filtering

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 commented:
I'm done with my refactorings. I won't make additional changes without re-review.

jessesimpson36 commented:
I rebased on latest main.

jessesimpson36 commented:
https://github.com/helm/helm/issues/13447#issuecomment-2725646423

linking my comment from the issue since it's relevant to this implementation

gjenkins8 commented:
> Here is my issue proposing additional fields to the ExecError type. [golang/go#71801](https://github.com/golang/go/issues/71801)

Nice!

gjenkins8 commented on pkg/engine/engine.go:
```diff
@@ -340,8 +363,70 @@ func cleanupExecError(filename string, err error) error {
+		}
+		if len(fileLocations) > 0 {
+			lastErr := fileLocations[len(fileLocations)-1]
+			if lastErr.message == traceable.message &&
```
Somewhat as a curiosity, this looks to de-duplicating a repeated error. If the go template engine has produced a repeated error, what is the reason we would try and elide that here?

jessesimpson36 replied:
I thought this was more of a golang quirk than something with just the templating engine.

In the unit test, where I have the error 
```
template: NestedHelperFunctions/templates/svc.yaml:1:9: executing "NestedHelperFunctions/templates/svc.yaml" at <include "nested_helper.name" .>: error calling include: template: NestedHelperFunctions/templates/_helpers_1.tpl:1:39: executing "nested_helper.name" at <include "common.names.get_name" .>: error calling include: template: NestedHelperFunctions/charts/common/templates/_helpers_2.tpl:1:50: executing "common.names.get_name" at <.Release.Name>: nil pointer evaluating interface {}.Name
```

If I call `err.Unwrap()`, I get the error with the same error mesage. it's only after calling it twice that I get the next meaningful block from the error message.  Which I thought, "why not call `err.Unwrap().Unwrap()`.  But I didn't want to call Unwrap twice because I can't guarantee that it will be a repeated error twice every time.

Perhaps the code upstream is calling `return fmt.Wrap(err)` with no meaningful changes to the error message somewhere.

jessesimpson36 replied:
https://github.com/helm/helm/pull/13586/commits/af956fdc3a326fc5f42cf7665a89b3aec357aa9a

I made this section more concise with this commit

jessesimpson36 commented on pkg/engine/engine.go:
```diff
@@ -340,8 +363,70 @@ func cleanupExecError(filename string, err error) error {
 	if len(parts) >= 2 {
 		return fmt.Errorf("execution error at (%s): %s", string(location), parts[1])
 	}
+	current := err
```
My understanding of golang errors is that it's like traversing a linked list. If you for whatever reason end up in a scenario where the error type is a circular list, then this would create an unbounded while-true loop.

But practically, that might just be an absurd possibility I'm trying to prevent.

jessesimpson36 replied:
I wrote some code in a test golang script to test out my idea of creating a loop with `errors.Unwrap` but it seems that golang is smart enough to seg-fault when you try to do this. I think I'm ready to change this to a while loop.

jessesimpson36 replied:
8b34ed3e77a21bc1ffeb0781bacae754ef62dda4

jessesimpson36 replied:
https://play.golang.com/p/5Vra__3FTdb

script where I test circular error loop

jessesimpson36 commented on pkg/engine/engine.go:
```diff
@@ -340,8 +363,70 @@ func cleanupExecError(filename string, err error) error {
+	fileLocations := []TraceableError{}
+	maxIterations := 100
+	for i := 0; i < maxIterations && current != nil; i++ {
+		if current == nil {
```
I must've had a while loop in here before instead of a for loop. I'll remove this block.

jessesimpson36 replied:
Done :+1: https://github.com/helm/helm/pull/13586/commits/199d9afe36c2593901f33548007a4f9cc11264fb

jessesimpson36 committed:
style: remove unnecessary break

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
refactor: use strings.Builder instead of string concatenation

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
refactor: define regexs at package scope

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 commented on pkg/engine/engine.go:
```diff
@@ -340,8 +363,70 @@ func cleanupExecError(filename string, err error) error {
+		if current == nil {
+			break
+		}
+		if i == maxIterations-1 {
```
I wouldn't say that it's redundant, as it does return instead of break.  If I removed it, then it would proceed with the next code after the for loop closes.  But it is confusing to have it like this.  why return? if you hit 100 iterations, then that probably means you're in an infinite loop, and whatever is at the end of the error line is going to be out-of-reach from the reformatting logic. In such a scenario, falling back to the old error message (unformatted by the code in this PR) is the safer approach.

That's the reasoning behind it, but this could be absurd. I'm willing to change it if this doesn't make sense to you.

jessesimpson36 replied:
from above:

> I wrote some code in a test golang script to test out my idea of creating a loop with errors.Unwrap but it seems that golang is smart enough to seg-fault when you try to do this. I think I'm ready to change this to a while loop.

https://github.com/helm/helm/commit/8b34ed3e77a21bc1ffeb0781bacae754ef62dda4

jessesimpson36 commented on pkg/engine/engine.go:
```diff
@@ -340,8 +363,70 @@ func cleanupExecError(filename string, err error) error {
+		current = errors.Unwrap(current)
+	}
+
+	finalErrorString := ""
```
Done! https://github.com/helm/helm/pull/13586/commits/88ef45362ee05c982895abebdf764ad4164ecaf0

jessesimpson36 commented on pkg/engine/engine.go:
```diff
@@ -340,8 +363,70 @@ func cleanupExecError(filename string, err error) error {
+		finalErrorString = finalErrorString + fileLocation.String()
+	}
+
+	if strings.TrimSpace(finalErrorString) == "" {
```
I can't think of a reason this would be empty. I will remove this.

jessesimpson36 committed:
refactor: remove impractical safety check

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 commented on pkg/engine/engine.go:
```diff
@@ -321,11 +321,34 @@ func cleanupParseError(filename string, err error) error {
+
+	execErrFmt, compileErr := regexp.Compile(`^template: (?P<templateName>(?U).+): executing (?P<functionName>(?U).+) at (?P<location>(?U).+): (?P<errMsg>(?U).+)(?P<nextErr>( template:.*)?)$`)
+	if compileErr != nil {
+		return err
```
I used MustCompile here

jessesimpson36 commented on pkg/engine/engine.go:
```diff
@@ -321,11 +321,34 @@ func cleanupParseError(filename string, err error) error {
+	}
+	execErrFmtWithoutTemplate, compileErr := regexp.Compile(`^template: (?P<templateName>(?U).+): (?P<errMsg>.*)(?P<nextErr>( template:.*)?)$`)
+	if compileErr != nil {
+		return err
```
I used MustCompile here, as you suggested :+1:

jessesimpson36 commented on pkg/engine/engine.go:
```diff
@@ -321,11 +321,34 @@ func cleanupParseError(filename string, err error) error {
+	// taken from https://cs.opensource.google/go/go/+/refs/tags/go1.23.6:src/text/template/exec.go;l=141
+	// > "template: %s: executing %q at <%s>: %s"
+
+	execErrFmt, compileErr := regexp.Compile(`^template: (?P<templateName>(?U).+): executing (?P<functionName>(?U).+) at (?P<location>(?U).+): (?P<errMsg>(?U).+)(?P<nextErr>( template:.*)?)$`)
```
done! https://github.com/helm/helm/pull/13586/commits/c836f85da214e327e70321de7698c8e567b35d2e

jessesimpson36 committed:
refactor: rename function and add doc-string

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 commented on pkg/engine/engine.go:
```diff
@@ -321,11 +321,34 @@ func cleanupParseError(filename string, err error) error {
+func (t TraceableError) String() string {
+	return t.location + "\n  " + t.executedFunction + "\n    " + t.message + "\n"
+}
 func cleanupExecError(filename string, err error) error {
```
Ok, I added this. Definitely let me know if you want it reworded. I'm unsure if what I wrote follows the same doc-string standards as the rest of the project. https://github.com/helm/helm/pull/13586/commits/a03ec15b135e9b3781077cae79fee2f045d2a981

jessesimpson36 commented:
I went through one round of implementing based on the feedback, but I would like a followup response on the topic of the unresolved comments (converting from for loop to unbounded while loop, error deduplication logic) before making any commits.

jessesimpson36 commented on pkg/engine/engine.go:
```diff
@@ -340,8 +363,70 @@ func cleanupExecError(filename string, err error) error {
+		var traceable TraceableError
+		if execErrFmt.MatchString(current.Error()) {
+			matches := execErrFmt.FindStringSubmatch(current.Error())
+			templateIndex := execErrFmt.SubexpIndex("templateName")
+			templateName := matches[templateIndex]
```
I missed this comment, somehow. I wonder if this was not added to the review? I can make this change.

jessesimpson36 committed:
refactor: shorten regex subexp syntax

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
refactor: replace if MatchString with FindStringSubMatch

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
refactor: switch to while loop instead of for to reduce unnecessary variables

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
refactor: use struct equality instead of comparing each composition

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 committed:
refactor: reduce flow-control operations

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 commented:
> I went through one round of implementing based on the feedback, but I would like a followup response on the topic of the unresolved comments (converting from for loop to unbounded while loop, error deduplication logic) before making any commits.

I think I addressed both of those topics now. Please see the messages in the "Review" threads.

jessesimpson36 commented:
@gjenkins8 may I have a re-review? (just a reminder in case this PR got lost in the noise)

jessesimpson36 commented:
rebased on main. Sorry for the extra noise this causes.

gjenkins8 reviewed (dismissed):
great, thanks for the updates

TerryHowe commented on pkg/engine/engine.go:
```diff
@@ -340,8 +364,42 @@ func cleanupExecError(filename string, err error) error {
+	}
 
-	return err
+	return fmt.Errorf("%s", finalErrorString.String())
```
```suggestion
	return errors.New(finalErrorString.String())
```

jessesimpson36 replied:
I added this in my latest commit.

TerryHowe reviewed (approved):
/lgtm with nit

jessesimpson36 committed:
fix: use errors.New instead of fmt.Errorf

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 commented:
@gjenkins8 your review got invalidated since my 2nd reviewer asked for a change. :sweat_smile: 
Could you re-review?

jessesimpson36 committed:
fix: address no-template-associated type of error

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 commented:
@scottrigby directed me towards a 3rd type of error possibility that looks like this:

> template: no template "nested_helper.name" associated with template "gotpl"

upstream it's found here:
https://cs.opensource.google/go/go/+/refs/tags/go1.23.6:src/text/template/exec.go;l=191

so I added a unit test and extra regex to match.

jessesimpson36 committed:
style: trim space from formatted error messages

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 commented:
what's interesting about that third type of error is that it's the only place in exec.go 
https://cs.opensource.google/go/go/+/refs/tags/go1.23.6:src/text/template/exec.go;l=191
that uses fmt.Errorf  instead of the usual state.errorf (the two regex's I covered).

jessesimpson36 committed:
test: use more realistic unit-test scenario by not relying on Release.Name

Signed-off-by: Jesse Simpson <jesse.simpson36@gmail.com>

jessesimpson36 commented:
I created this repo to help with  manually testing helm in different error scenarios.

https://github.com/jessesimpson36/helm_error_charts/tree/main

in the divide_by_zero test chart, I found that this error message is not formatted. I'm not sure yet why.

jessesimpson36 commented:
On the helm dev call today,  I brought up this issue:
https://github.com/helm/helm/issues/30857

which is mildly related to this PR, since this PR will add newline characters to this debug log line.  The outcome of that discussion was that the extra debug log line should not be there, so I will not put any effort to re-shaping that error message to look better.  I'll open a separate PR for removing that debug log line, since it is an independent change from formatting the error message.

jessesimpson36 commented:
> I created this repo to help with manually testing helm in different error scenarios.
> 
> https://github.com/jessesimpson36/helm_error_charts/tree/main
> 
> in the divide_by_zero test chart, I found that this error message is not formatted. I'm not sure yet why.

I now know why the divide_by_zero chart returned an unformatted error. The reason is that the final error in the chain does not match any of the regex's.  I previously thought that I had covered all of the error scenarios in the text/template's exec.go, because I covered all of the occurrences of `(*state).errorf()` and `fmt.Errorf` function calls.  Where I was wrong is that

https://cs.opensource.google/go/go/+/refs/tags/go1.23.6:src/text/template/exec.go;l=816

In this line, `%w` is used, which wraps the `err` object from the `safeCall` caught-panic above (on 811). So when we do `current = current.Error()`, we get a new error object with the text "runtime error: integer divide by zero" which doesn't match the regex's.

This is the only error in the file that makes use of `%w`, so if we handle this case, we should be good. I'm also not sure how common this error pathway would be.  It's not super painful for users if a small amount of errors are unformatted.  But what I think would be ideal is :

if the error matches `error calling %s: %w`, we format that chunk as though `%w` was just part of the error string. Then, break out of the `current = current.Error()` loop, so that it doesn't try to match `%w` .

Does that sound like a good plan? or would you rather that be done in a separate branch or separate PR given this one already has 2 reviews?

jessesimpson36 commented:
ermmm... that idea didn't work out well.

`error calling %s: %w` is not at any point in an error string on its own. its appended after 
`template: %s: executing %q at <%s>: %s` on the same error struct. and concatenating the 2 regex's or doing a check for "error calling" after matching `template: %s: executing %q at <%s>: %s` doesn't make sense because those are common error message words. :(

jessesimpson36 commented:
This commit seems to work https://github.com/jessesimpson36/helm/pull/1/commits/a94643923c30fb664bfb3a3327708ded882a7cac

I'm testing it in a different branch, but I'm not as confident about it not having whacky side-effects compared to the idea of simply accepting some error messages will be unformatted.

jessesimpson36 commented:
I wanted to see how prevalent the code-path that triggers `divide_by_zero` would be. Could I trigger an error that results in an unformatted string with another command other than `div` . To do this, I inserted a `fmt.Println("Calling: " + name)` just above the safeCall function in text/template's exec.go. Then I ran our unit tests, and grepped for that print statement (and removed duplicates). the results are my candidates:

```
Calling: default
Calling: dict
Calling: div
Calling: exclaim
Calling: fail
Calling: fromJson
Calling: fromJsonArray
Calling: fromToml
Calling: fromYaml
Calling: fromYamlArray
Calling: Get
Calling: getHostByName
Calling: include
Calling: indent
Calling: len
Calling: lookup
Calling: lower
Calling: merge
Calling: printf
Calling: quote
Calling: required
Calling: title
Calling: toJson
Calling: toToml
Calling: toYaml
Calling: toYamlPretty
Calling: tpl
Calling: upper

```

I went through and manually tested each of these commands to trigger an error to see if the error was formatted or not. Many of these commands do not panic with an error, and instead will opt to return an empty string if the command is invalid. Some commands will return a 1 line short error string, like in the case of "required" or "fail". I think that's fine to not print the entire error path for those situations, since a helm maintainer knows where you put fail statements in your own helm chart. I tried  using invalid json and invalid merge calls and those were successfully formatted.  I did not see any scenarios like  `div 5 0` producing an unformatted error.

scottrigby reviewed (approved):
Thanks for all the hard work on this. LGTM

TerryHowe commented:
Yeh, nice one
